### PR TITLE
Allow arbitrary --user values (mostly)

### LINKED ISSUE #46: Using postgresql with external volume from Mac
I am trying to mount it directly from the Mac file system.

``` bash
docker run  -v $HOME/postgres_data:/var/lib/postgresql kartoza/postgi
```

This results in a bunch of errors about user permissions:

```
initdb: could not create directory "/var/lib/postgresql/9.3/main/global": Permission denied
2015-01-07 16:33:52 UTC FATAL:  data directory "/var/lib/postgresql/9.3/main" has wrong ownership
2015-01-07 16:33:52 UTC HINT:  The server must be started by the user that owns the data directory.
```

Can a directory shared from Mac file system ever have the read/write permissions that postgres apparently needs?

### LINKED ISSUE #116: chmod: changing permissions of 'var/lib/postgresql/data': Permission denied
I have slightly extended the official postgres Image in the following way (only the COPY is really important, to do some stuff to use SonarQube with Postgres):
# Version: 1.0.0

FROM postgres:9.5
MAINTAINER Wolfgang Grossinger "wolfgang.grossinger@gmail.com"
COPY sonarqube.sql /docker-entrypoint-initdb.d
RUN localedef -i en_US -c -f UTF-8 -A /usr/share/locale/locale.alias en_US.UTF-8
ENV LANG en_US.utf8
ENV POSTGRES_PASSWORD=postgres
ENV POSTGRES_USER=postgres
RUN chmod 777 /var/lib/postgresql/data

After building the image I try to start it with the following command:

sudo docker run --name postgres_sonar -p 5432:5432 -v /data:/var/lib/postgresql/data wogro/postgres_sonar

I get the following error message: 

chmod: changing permissions of ‘/var/lib/postgresql/data’: Permission denied

I don't know what the reason is (probably it's me), but it is something i would never expect from an official image (I just would expect that mounting a volume works :-))

### LINKED ISSUE #206: How to change the gid and uid of user account postgres?
Hi,
Currently we are using postgres image in our Kubernetes environment, and we use nfs as our data volume solution, that is to say, `/var/lib/postgresql/data` in postgres image will be mounted to a share folder on nfs server finally, let's say `/var/vols/data` on nfs server.
The problem is in nfs server configuration, administrator sets the `anongid` and `anonuid` to an id that is different with postgres account in docker image, for example, 1999, while the gid and uid of `postgres` account is 999, and then we got this error:
`chown: changing ownership of ‘/var/lib/postgresql/data’: Operation not permitted`

Any suggested solutions to solve this problem?
Thanks,
Viciss

### LINKED ISSUE #251: uid collision on Centos 7 host
The hard-coded uid=999 value conflicts with the systemd-bus-proxy user on CentOS 7
```
$ grep :999: /etc/passwd
systemd-bus-proxy:x:999:997:systemd Bus Proxy:/:/sbin/nologin
```

This makes it difficult to use a volume mapped to a host directory as it is owned by the above user.   Is there a recommended way to handle this?

### PR DESCRIPTION
**Pull request by tianon**
One special case is that `initdb` _requires_ the current user to exist in `/etc/passwd`, but running PostgreSQL itself does not require that.

As discussed over in https://github.com/docker-library/postgres/pull/93#issuecomment-273877372.

See also docker-library/rabbitmq#60, docker-library/cassandra#48, docker-library/mongo#81, docker-library/redis#48, docker-library/mysql#161, docker-library/mariadb#59, docker-library/percona#21, and https://github.com/docker-library/ghost/pull/54.

Closes #46
Closes #116
Closes #206
Closes #251
Ref #28

### CODE REVIEW

**Comment by tianon** [comment]
```console
$ docker run -it --rm --user 1000:1000 postgres
initdb: could not look up effective user ID 1000: user does not exist

$ docker run -it --rm --user www-data postgres
The files belonging to this database system will be owned by user "www-data".
This user must also own the server process.
...

$ docker run -it --rm --user "$(id -u):$(id -g)" -v /etc/passwd:/etc/passwd:ro postgres
The files belonging to this database system will be owned by user "tianon".
This user must also own the server process.
...

$ dir="$(mktemp -d)"
$ docker run -it --rm -v "$dir":/var/lib/postgresql/data postgres
... (let initialization finish, then stop the server)
$ sudo chown -R 1000:1000 "$dir"
$ docker run -it --rm -v "$dir":/var/lib/postgresql/data --user 1000:1000 postgres
LOG:  database system was shut down at 2017-01-19 23:06:31 UTC
LOG:  MultiXact member wraparound protections are now enabled
LOG:  database system is ready to accept connections
LOG:  autovacuum launcher started
```

**Comment by yosifkit** [comment]
With regards to my comment of it breaking people that are taking advantage of being `root` while in `initdb.d` scripts, I would point out that the change was pushed to `mysql` on [May 17, 2016](https://github.com/docker-library/official-images/pull/1740) and to `mariadb` and `percona` on [May 18. 2016](https://github.com/docker-library/official-images/pull/1753) and there has yet to be an issue reported about it.

**Comment by yosifkit** [comment]
We'll probably want to add some docs about the limitations observed here.

**Comment by tianon** [comment]
https://github.com/docker-library/docs/pull/802 :+1:

**Comment by shane-axiom** [comment]
FWIW this __did__ cause an issue for us, as a `.pgpass` file provided in a volume mounted to `/root/.pgpass` suddenly stopped working. It took a while to figure out why with inconsistent behavior between various `postgres:9.6` images cached on different workstations/servers. Easy fix once we figured out what was going on, but it wasn't immediately obvious.

The __How to extend this image__ section in the docs stll doesn't explicitly say that `*.sh` scripts will be run by the `postgres` user by default, that might help.

**Comment by tianon** [comment]
Excellent idea, thanks @shane-axiom :heart:

I've filed a PR at https://github.com/docker-library/docs/pull/848 :+1:

**Comment by willemavjc** [comment]
> The following comment from [docker-library/mysql#161 (comment)](https://github.com/docker-library/mysql/pull/161#issuecomment-212134139) also applies here:
> 
> > I think the main difference is that the `initdb.d` scripts will no longer be run as root; not sure if that will break too many users.

In fact I think it does ...for the entire Windows community. 

"root" is granted a useful privilege when Docker is allowed to share Windows drives (eg. C:): To be able to write onto the Windows (host) filesystem when mapping directories through the -v option; -v option which does *not* support any mount options from Docker for recall. (See [Docker Documentation](https://docs.docker.com/engine/reference/commandline/volume_create/) "The built-in local driver on Windows does not support any options.")

Because "root" is no longer in the party, Windows users are no longer able to remap postgres' data and that is what a large part of them is complaining about since 3 years now.

Demonstration:
```
docker run -it --rm --user root -v D:/Docker/postgres:/var/lib/postgresql/data postgres:11.2-alpine
The files belonging to this database system will be owned by user "postgres".
...
running bootstrap script ... 2019-03-24 17:22:58.583 UTC [47] FATAL:  data directory "/var/lib/postgresql/data" has wrong ownership
2019-03-24 17:22:58.583 UTC [47] HINT:  The server must be started by the user that owns the data directory.
```

We can see that "root" user was just ignored resulting in "postgres" default user and that so known wrong ownership issue.

But if you start that very same image with a sh, the freshly created container will be able to write, modify and delete what you want on that directory.

```
docker container run --rm -it -v D:/Docker/postgres:/var/lib/postgresql/data postgres:11.2-alpine sh
$ cd /var/lib/postgresql/data
$ touch test
$ ls
```

Not a single line of the postgres' documentation about users and drives mapping is true for a Windows environment.

The only way I did achieve was to pass through a data container, the old way in the end.

**Comment by tianon** [comment]
This PR isn't (and really can't) be the cause of PostgreSQL not working on Windows -- it doesn't change anything about how `postgres` itself runs.  The bit you quoted refers to how `/docker-entrypoint-initdb.d` _scripts_ are run -- previously those ran as root, now they run as the same user PostgreSQL does (which makes way more sense anyhow).

If you want to bypass our behavior entirely, it should be trivial to do so via `--entrypoint postgres` (which would be the only way to start the `postgres` daemon itself as `root` with the current image script, although I'm not sure whether PostgreSQL itself will allow that or whether it balks at running as root).

**Comment by willemavjc** [comment]
Mmh, not sure to understand what is going on now:

```
docker run -it --user root -v D:/Docker/postgres:/var/lib/postgresql/data --entrypoint="" postgres:11.2-alpine sh
/var/lib/postgresql # pg_ctl start
pg_ctl: cannot be run as root
Please log in (using, e.g., "su") as the (unprivileged) user that will
own the server process.
/var/lib/postgresql # su postgres
~ $ pg_ctl start
pg_ctl: directory "/var/lib/postgresql/data" is not a database cluster directory
~ $ ls -la /var/lib/postgresql/
total 16
drwxr-xr-x    1 postgres postgres      4096 Mar 26 16:16 .
drwxr-xr-x    1 root     root          4096 Mar  8 02:22 ..
-rw-------    1 postgres postgres        41 Mar 26 16:19 .ash_history
drwxrwxrwx    2 root     root          4096 Mar 26 16:19 data
~ $
```

**Comment by yosifkit** [comment]
@willemavjc, try this comment: https://github.com/docker-library/postgres/issues/558#issuecomment-472234418. 

tldr: it is not possible on Docker for Windows for Postgres to use a shared folder to the host

**Comment by willemavjc** [comment]
> @willemavjc, try this comment: [#558 (comment)](https://github.com/docker-library/postgres/issues/558#issuecomment-472234418).
> 
> tldr: it is not possible on Docker for Windows for Postgres to use a shared folder to the host

Thanks for sharing some links. Yes, that's the conclusion I ended up with. And yes, the only way with Docker for Windows to make "persistent" data is through a volume which is well documented on Docker documentation; in fact that was the first thing I tried when discovering docker/postgres.

BUT... As a matter of fact, no one no longer can call this "persistent" data, "semi-persistent" data at the most if one really want to believe his data will last for ...some time.

Because the Docker volume is on Windows limited to an inner directory within Moby, then the data will not survive any issue with Docker itself requiring a reset, for whatever the reason may be: unrecoverable crashes, environment refresh, etc. Said differently, the data is not onto the host fs, living on its own.

Trying to fake the volume within Moby with a symlink made me realize the things which dysfunction: Moby writes on shared volumes as root, not a unprivileged user. When using Alpine of Postgres, one can write to Windows fs because I believe it passes all the way down root privileges. On the other hand, when logged as postgres, the story ends fast because it simply is not privileged; --privileged=true did nothing by the way.

So the solution would either manage somehow postgres to write data as root or Docker to remap unprivilege writes operations as root for shared folders.

Opened for any discussion on this point of view.
